File: C++/프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.cpp
#include <string>
#include <vector>
#include <cctype>
#include <algorithm>

using namespace std;

int solution(string str1, string str2) {
    vector<string> vec1, vec2;
    transform(str1.begin(), str1.end(), str1.begin(), ::tolower);
    transform(str2.begin(), str2.end(), str2.begin(), ::tolower);
    
    for(int i = 1; i < max(str1.size(), str2.size()); i++) {
        if(i < str1.size() && 0 < isalpha(str1[i - 1]) && 0 < isalpha(str1[i])) {
            vec1.push_back(string{str1[i - 1]} + str1[i]);
        }
        if(i < str2.size() && 0 < isalpha(str2[i - 1]) && 0 < isalpha(str2[i])) {
            vec2.push_back(string{str2[i - 1]} + str2[i]);
        }
    }
    sort(vec1.begin(), vec1.end());
    sort(vec2.begin(), vec2.end());
    int index1 = 0, index2 = 0;
    
    int intersection = 0;
    while(index1 < vec1.size() && index2 < vec2.size()) {
        if(vec1[index1] == vec2[index2]) {
            intersection++;
            index1++; index2++;
        } else if(vec1[index1] < vec2[index2]) {
            index1++;
        } else if(vec1[index1] > vec2[index2]) {
            index2++;
        }
    }
    
    
    float unionn = vec1.size() + vec2.size() - intersection;
    // cout << "intersection: " << intersection << endl;
    // cout << "union: " << unionn << endl;
    return ((unionn == 0) ? 1.0 : intersection / unionn) * 65536 / 1;
}

File: Python3/프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    answer = 0
    def divide_list(dstr):
        dlist = list()
        
        before = None
        for dc in dstr:
            dc = dc.lower()
            if before and before.isalpha() and dc.isalpha(): #첫번째 인덱스가 아닌 경우
                dlist.append(str(before + dc))
            before = dc
        
        return sorted(dlist)
    
    
    list1, list2 = divide_list(str1), divide_list(str2)
    # print(list1)
    # print(list2)
    
    #교집합 원소 찾기
    index1, index2 = 0, 0
    while index1 < len(list1) and index2 < len(list2):
        if list1[index1] == list2[index2]:
            index1 += 1
            index2 += 1
            answer += 1
        elif list1[index1] < list2[index2]:
            index1 += 1
        elif list1[index1] > list2[index2]:
            index2 += 1
    
    answer1, answer2 = answer, len(list1) + len(list2) - answer
    # print("교집합 원소 수:", answer1)
    # print("합집합 원소 수:", answer2)
    
    return ((answer1 / answer2) if answer2 > 0 else 1) * 65536 // 1